types/styled-components/index.d.ts

```diff
@@ -48,185 +67,324 @@ export interface Styles {
 
 export type InterpolationFunction<P> = (props: P) => Interpolation<P>;
 
-type Attrs<P, A extends Partial<P>, T> = {
+type Attrs<P, A extends Partial<P>, T> =
+    | ((props: ThemedStyledProps<P, T>) => A)
+    | A;
+type DeprecatedAttrs<P, A extends Partial<P>, T> = {
     [K in keyof A]: ((props: ThemedStyledProps<P, T>) => A[K]) | A[K]
 };
 
 export type ThemedGlobalStyledClassProps<P, T> = P & {
+    suppressMultiMountWarning?: boolean;
     theme?: T;
 };
 
-export interface GlobalStyleClass<P, T>
-    extends React.ComponentClass<ThemedGlobalStyledClassProps<P, T>> {}
+export interface GlobalStyleComponent<P, T>
+    extends React.ComponentClass<ThemedGlobalStyledClassProps<P, T>> { }
 
-export interface StyledComponentClass<P, T, O = P>
-    extends React.ComponentClass<ThemedOuterStyledProps<O, T>> {
-    withComponent<K extends keyof JSX.IntrinsicElements>(
-        tag: K,
-    ): StyledComponentClass<
-        JSX.IntrinsicElements[K],
+// remove the call signature from StyledComponent so Interpolation can still infer InterpolationFunction
+type StyledComponentInterpolation = Pick<
+    StyledComponent<any, any>,
+    keyof StyledComponent<any, any>
+    >;
+
+// abuse Pick to strip the call signature from ForwardRefExoticComponent
+type ForwardRefExoticBase<P> = Pick<
+    React.ForwardRefExoticComponent<P>,
+    keyof React.ForwardRefExoticComponent<any>
+    >;
+
+// any doesn't count as assignable to never in the extends clause, and we default A to never
+type AnyStyledComponent =
+    | StyledComponent<any, any, any, any>
+    | StyledComponent<any, any, any>;
+
+export interface StyledComponent<
+    C extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
+    T extends object,
+    O extends object = {},
+    A extends keyof any = never
+    > extends ForwardRefExoticBase<StyledComponentProps<C, T, O, A>> {
+    // add our own fake call signature to implement the polymorphic 'as' prop
+    // NOTE: TS <3.2 will refuse to infer the generic and this component becomes impossible to use in JSX
+    // just the presence of the overload is enough to break JSX
+    //
+    // TODO (TypeScript 3.2): actually makes the 'as' prop polymorphic
+    // (
+    //     props: StyledComponentProps<C, T, O, A> & { as?: never }
+    //   ): React.ReactElement<StyledComponentProps<C, T, O, A>>
+    // <AsC extends keyof JSX.IntrinsicElements | React.ComponentType<any> = C>(
+    //   props: StyledComponentPropsWithAs<AsC, T, O, A>
+    // ): React.ReactElement<StyledComponentPropsWithAs<AsC, T, O, A>>
```

Comment:
Good news is, these two overloads are sufficient to implement the `as` prop in TS 3.2. The bad news is, their presence at all causes everything to break in TS<=3.1.